R/archive.R

```diff
@@ -377,6 +354,63 @@ new_epi_archive <- function(
   )
 }
 
+#' given a tibble as would be found in an epi_archive, remove duplicate entries.
+#' @description
+#' works by shifting all rows except the version, then comparing values to see
+#'   if they've changed. We need to arrange in descending order, but note that
+#'   we don't need to group, since at least one column other than version has
+#'   changed, and so is kept.
+#' @keywords internal
+#' @importFrom dplyr filter
+compactify_tibble <- function(df, keys) {
+  df %>%
+    arrange(!!!keys) %>%
+    filter(if_any(
+      c(everything(), -version), # all non-version columns
+      ~ !is_locf(.)
+    ))
+}
+
+#' get the entries that `compactify_tibble` would remove
+#' @keywords internal
+#' @importFrom dplyr filter if_all everything
+removed_by_compactify <- function(df, keys) {
+  df %>%
+    arrange(!!!keys) %>%
+    filter(if_all(
+      c(everything(), -version),
+      ~ is_locf(.)
+    )) # nolint: object_usage_linter
+}
+
+#' Checks to see if a value in a vector is LOCF
+#' @description
+#' LOCF meaning last observation carried forward. lags the vector by 1, then
+#'   compares with itself. For floats it uses float comparison, otherwise it
+#'   uses equality.
+#' `NA`'s are considered equal to `NA`'s, while nan's are not.
+#' @importFrom dplyr lag if_else near
+#' @keywords internal
+is_locf <- function(vec) { # nolint: object_usage_linter
+  lag_vec <- lag(vec)
+  if (typeof(vec) == "double") {
+    if_else(
+      !is.na(vec) & !is.na(lag_vec),
+      near(vec, lag_vec),
```

Comment:
Given that  `R` only has absolute tolerance:
```R
near(1e-10, 1e-10)
[1] TRUE
near(1e-10, 1.1e-10)
[1] TRUE
```
we should definitely include a `compactify_tol argument, and mostly leave it undiscussed, since I seriously doubt most relevant quantities are going to be ~1e-10.

> issue?: we can get different results from downloading....

Floating point is inherently messy. Give people the option to set the precision and call it a day. It's only ever a problem if your values are near the tolerance, which, kind of by definition, means you set the tolerance to high.

Reply:
Random notes: `dplyr::near` only has abs tol.  `all.equal` uses relative with probably an absolute backup.

Tol parm sounds good.  Should also be an arg to `revision_summary` and used when re-compactifying.

The 1e-10 (or default ~1.5e-8).... was thinking maybe we could get near it if using basic rates rather than per 100k population, but actually running numbers doesn't seem likely.  Not sure about what scale NWSS concentrations though.  Anyway, having a parm would allow us to avoid issues here anyway.

Reply:
oh right, I ran across near before `all.equal` when trying to figure this out. Otoh,
```r
all.equal(1e-10,1e-10)
[1] TRUE
> all.equal(1e-10,1.1e-10)
[1] TRUE
```
so I'm not sure it really matters.
NWSS ranges from ~1e7 to 1e14, so it would pose some difficulty for any absolute tolerance scheme (not for a single value mind you, but across locations). The spurious revisions are getting taken care of in python though, so we don't really need to worry about it too much.